service/matching/matching_engine.go

```diff
@@ -837,6 +837,155 @@ func (e *matchingEngineImpl) listTaskQueuePartitions(request *matchingservice.Li
+				updatedClock,
+				data.GetVersioningData(),
+				req.GetReplaceAssignmentRule(),
+			)
+			if err != nil {
+				// operation can't be completed due to limits. no action, do not replicate, report error
+				return nil, false, err
+			}
+		case *workflowservice.UpdateWorkerVersioningRulesRequest_DeleteAssignmentRule:
+			versioningData, err = DeleteAssignmentRule(
+				updatedClock,
+				data.GetVersioningData(),
+				req.GetDeleteAssignmentRule(),
+			)
+			if err != nil {
+				// operation can't be completed due to limits. no action, do not replicate, report error
+				return nil, false, err
+			}
+			if ns.ReplicationPolicy() == namespace.ReplicationPolicyMultiCluster {
+				operationCreatedTombstones = true
+			} else { // todo: figure out if this is the right thing to do with tombstones
+				// We don't need to keep the tombstones around if we're not replicating them.
+				versioningData = ClearTombstones(versioningData)
+			}
+		case *workflowservice.UpdateWorkerVersioningRulesRequest_InsertCompatibleRedirectRule:
+			versioningData, err = InsertCompatibleRedirectRule(
+				updatedClock,
+				data.GetVersioningData(),
+				req.GetInsertCompatibleRedirectRule(),
+				e.config.VersionCompatibleRedirectRuleLimitPerQueue(ns.Name().String()),
+			)
+			if err != nil {
+				// operation can't be completed due to limits. no action, do not replicate, report error
+				return nil, false, err
+			}
+		case *workflowservice.UpdateWorkerVersioningRulesRequest_ReplaceCompatibleRedirectRule:
+			versioningData, err = ReplaceCompatibleRedirectRule(
+				updatedClock,
+				data.GetVersioningData(),
+				req.GetReplaceCompatibleRedirectRule(),
+			)
+			if err != nil {
+				// operation can't be completed due to limits. no action, do not replicate, report error
+				return nil, false, err
+			}
+		case *workflowservice.UpdateWorkerVersioningRulesRequest_DeleteCompatibleRedirectRule:
+			versioningData, err = DeleteCompatibleRedirectRule(
+				updatedClock,
+				data.GetVersioningData(),
+				req.GetDeleteCompatibleRedirectRule(),
+			)
+			if err != nil {
+				// operation can't be completed due to limits. no action, do not replicate, report error
+				return nil, false, err
+			}
+			if ns.ReplicationPolicy() == namespace.ReplicationPolicyMultiCluster {
+				operationCreatedTombstones = true
+			} else { // todo: figure out if this is the right thing to do with tombstones
+				// We don't need to keep the tombstones around if we're not replicating them.
+				versioningData = ClearTombstones(versioningData)
+			}
+		case *workflowservice.UpdateWorkerVersioningRulesRequest_CommitBuildId_:
+			// todo: not sure here
+		}
+		// Avoid mutation
+		ret := common.CloneProto(data)
+		ret.Clock = updatedClock
+		ret.VersioningData = versioningData
+		return ret, true, nil
+	})
+
+	if err != nil {
+		return nil, err
+	}
+
+	// todo: figure out correct tombstone behavior
+	// Only clear tombstones after they have been replicated.
+	if operationCreatedTombstones {
```

Comment:
shouldn't we call UpdateUserData for all successful operations to persist new changes?

Reply:
the whole big switch statement is inside a call to `UpdateUserData`, so we are calling it